aixPB: first PR to `start` the `cleanup and restructure` of AIX playbooks

aixtools opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.
-->

##### Checklist
<!-- For completed items, change [ ] to [x]. Leave unchecked if not required -->

- [x] commit message has one of the [standard prefixes](https://github.com/adoptium/infrastructure/blob/master/CONTRIBUTING.md#commit-messages)
- [ ] [faq.md](https://github.com/adoptium/infrastructure/blob/master/FAQ.md) updated if appropriate
- [ ] other documentation is changed or added (if applicable)
- [ ] playbook changes run through [VPC](https://ci.adoptopenjdk.net/view/Tooling/job/VagrantPlaybookCheck/) or [QPC](https://ci.adoptopenjdk.net/view/Tooling/job/QEMUPlaybookCheck/) (if you have access)
- [ ] for inventory.yml changes, bastillion/nagios/jenkins updated accordingly

This PR is the first of several to:
fixes #1582 
fixes #1918

sxa reviewed (changes requested):
Thanks for this - took a while to review but there are a lot of individual points in here! One suggestion particuarly when putting in large changes like this - put a summary of the "philiosophies" of what you're doing in the description of the issue to make it easy for people to understand what the PR is intending to do overall.

Main themes:
- Security role is good and I want to get something similar rolled out to the other OS playbooks in the near future too
- I'm not convinced about the use of a separate user role for the reasons in the comments. as that diverges from how the other OS playbooks do things.
- I don't want to remove the ability for the playbooks to install openssl (Still required for OpenSSL I think) and the JDK7 boot JDK. The default execution of the playbooks without any parameters needs to be the one that sets up the machines "correctly" for the adoptium build/test farm. The use of `/Vendor_Files` isn't ideal, but it needs to include this stuff by default and not risk leaving machines "not fit for purpose"

Otherwise these all look like good enhancements, so thanks for putting the work in on this.

aixtools commented on ansible/playbooks/AdoptOpenJDK_AIX_Playbook/main.yml:
```diff
@@ -234,26 +250,26 @@
       shell: |
         swap_line=$(lsps -a | tail -n 1)
-        curr_swap_size=$(echo $swap_line | awk '{print $4}' | awk -FM '{print $1}')
+        curr_size=$(echo $swap_line | awk '{print $4}' | awk -FM '{print $1}')
```
Line length under 80, iirc.

aixtools commented on ansible/playbooks/AdoptOpenJDK_AIX_Playbook/main.yml:
```diff
@@ -74,13 +60,40 @@
+      tags:
+        - adoptopenjdk
+        - users
+    - role: users
```
- Actually, what I am trying to achieve here is being able to specify a username (e.g., `aixtools` as username) on the command-line using `ansible-playbook --tags=users -e "users=aixtools" ..../main.yml` because my local user for building needs the settings `jenkins` needs.
- what I need to add to the documentation (when I learn where it is to go, regular FAQ.md, or a special one) - add the need for: `touch roles/users/tasks/aixtools.yml`
- I'll add more lower-level tags such as:
```
./tasks/jenkins.yml:
- name: warn about failed copy
  debug:
    msg: "{{ Jenkins_User_SSHKey }} failed to copy"
  when: jenkins_rsa.failed == true
  tags: jenkins_user
- Last, considering the instructions is FAQ.md:
`ansible-playbook -b -i inventory_file --skip-tags adoptopenjdk,jenkins_user playbooks/AdoptOpenJDK_Unix_Playbook/main.yml`
The tag `adoptopenjdk` before the on the users role - skips all the users creations - including jenkins.

aixtools replied:
> Hmmm I'm going to need to be convinced on this one. Three reasons:
> 
>     * This makes the process inconsistent between 
a) Maybe the only things *nix's have in common is _nix_ in the branding.
b) the bigger issue is that we are still in the process of transforming the flatfile `aix.yml` into proper ansible structure(s).
c) copy/paste a bunch of lines from `aix.yml' to roles/somerolename/tasks/main.yml does not make it a _proper_ role.
> AIX and other UNIX platforms which makes maintenance harder so I wouldn't want to accept this unless it was proposed across all OS playbooks (same for the use of the `adoptium` group)
> 
- removed the adoptium group. As I mentioned above - this is a `security` consideration to not have everything in a default group. However, as adoptium internal systems are not really commercial production systems - I'll skip it. If/when this were to be run on a non-Ansible project system I would consider the group more important.
- as to unfying the playbooks - I have glanced at the *Unix* playbooks and see many opportunities for making better use of Ansible constructs AND I see the *AIX* playbooks as a relatively safe way to gain experience with using code such as:
```
- name: Include Platform specific variables
  include_vars: "{{ansible_os_family}}".yml"
- name: include Platform specific configuration tasks
  include_tasks: "config-{{ ansible-os-family }}.yml"
```
And in a proper role these would be located in `roles/rolename/{vars|tasks}/` respectively.
- imho: having all the users in one area is cleaner, and easier to maintain. And a proper role (user management, versus task after task after task to do, or not do users.
-  I cannot speak re: the *Unix* users, because I have not looked at them, but with the AIX users one change - especially with ulimits - was three edits.
- re: ease of maintaining users: it took a few passes to get all the common material centralized, ie.e., what was common, what was unique was not something jumping out at you.
- having all user information in one area (directory) makes, imho, triage of user issues much easier.
- ansible has it's idiosyncrasies - and where I had the most difficulty was finding a good way to get _loops_ and _tags_ to work.
>     * I think there's enough diffferences between the users that this is best kept as separate roles. The requirements for each user are fairly different and the `zeus` user in particular should be as basic/clean as possible.
- In AIX - ulimits are ulimits, don't care what your username is.
- The 'customization' for zeus, if you want that is to remove the non-ulimits attributes.
- The nagios 'user' confused me, but I left it there - **note** on the adopt nodes only three of the 8 have a authorized_key that exists AND is not zero bytes.
- In any case, I feel it is much more `ansible way` to break things into individual playbooks to perform one task
- one book over ansible gives (design) advice to keep individual files to under 100 lines - as being more modular
- ansible `roles` are not merely a way to move tasks out of ./main.yml to a sub-directory, but to organize defaults and vars in the role.
- (I spent a lot of time researching Ansible to come up with this model).
> 
>     * Most of the special options that we add to the users for various permissions are only required on the `jenkins` user so unless there's a reason to I wouldn't want to add them to the others (I think these roles are applying a lot of the settings to all of them)
# Anyway
- awaiting your reply.

aixtools commented on ansible/playbooks/AdoptOpenJDK_AIX_Playbook/roles/rbac/tasks/main.yml:
```diff
@@ -13,33 +13,37 @@
   shell: lsauth ojdk.rtclk
   register: rtclk_exists
-  ignore_errors: yes
+  failed_when: rtclk_exists.rc != 0 and rtclk_exists.rc != 2
```
- I have added/changed comments to add the second issue. The first one was already there.
- Note also: you have never given permission to test RBAC works - you pushed that we rely on undocumented `capabilities`. As such, while the RBAC settings are added to commands, no user is assigned a role to make use of them.

aixtools commented:
@sxa - made several replies, and some changes. May I suggest a call as a more efficient way to discuss what works and doesn't work? I'll add a summary here when we get in sync.

aixtools commented on ansible/playbooks/AdoptOpenJDK_AIX_Playbook/main.yml:
```diff
@@ -6,16 +6,23 @@
+  pre_tasks:
+    - name: Make Adoptium group
+      group:
+        name: adoptium
```
- it was a security thought. Common practice is to not use the group `staff`
- I'll remove for now.

aixtools commented on ansible/playbooks/AdoptOpenJDK_AIX_Playbook/roles/openssl/tasks/main.yml:
```diff
@@ -1,40 +1,29 @@
-    dest: /tmp/openssl-1.0.2.1601.tar.Z
-  when: does_openssl_exist.rc != 0
-  tags: openssl
+- name: "Fail: OpenSSL 1.0.2p or later required"
```
Hmm, where di my earlier reply go...
- OpenSSL is an OS supported file
- this was likely important in 2017 - as the default for AIX 7.1 TL4 SP4 was openssl.base.1.0.2.800
- imho, no longer important in 2022 - with current levels at 1.0.2.2102 and 1.1.1.1200
- Should not be the responsibility of the _Ansible master_. It is the responsibility of the PAAS/IAAS System admin) the hosting virtual machoine

sxa replied:
Yep i'd agree with all of that. I'll make a suggested change to clarify that in the error message.

aixtools commented:
> Thanks for this - took a while to review but there are a lot of individual points in here! One suggestion particuarly when putting in large changes like this - put a summary of the "philiosophies" of what you're doing in the description of the issue to make it easy for people to understand what the PR is intending to do overall.
> 
> Main themes:
> 
>     * Security role is good and I want to get something similar rolled out to the other OS playbooks in the near future too
> 
>     * I'm not convinced about the use of a separate user role for the reasons in the comments. as that diverges from how the other OS playbooks do things.
> 
>     * I don't want to remove the ability for the playbooks to install openssl (Still required for OpenSSL I think) and the JDK7 boot JDK. The default execution of the playbooks without any parameters needs to be the one that sets up the machines "correctly" for the adoptium build/test farm. The use of `/Vendor_Files` isn't ideal, but it needs to include this stuff by default and not risk leaving machines "not fit for purpose"
> 
> 
> Otherwise these all look like good enhancements, so thanks for putting the work in on this.

Thx for the summary.

Another point: I paused here with reorganization to let us get in sync. I have a next iteration with bootjdk roles streamlined.

As to philosophy - minimize logic in main.yml - less comments I would like to see it get to 50 to 70 lines max. Less comments it is nearly 290 lines now. Additional cleanup will come with migrating tasks in main.yml into respective roles.

However, in a word: cleanup, in three words: cleanup and document

aixtools commented:
_Adding a extra comment because my experience with other github projects is that a comment at this level is needed to make sure the comments to requests for change are 'announced'.
At least that was 'their' excuse for not replying.
Please - accept my apology (when not true re: notifications) and disregard._

smlambert commented:
Noting that I believe Stewart is on vacation until Monday - so response may be a bit.

sxa commented on ansible/playbooks/AdoptOpenJDK_AIX_Playbook/roles/bootjdk7/tasks/main.yml:
```diff
@@ -2,16 +2,32 @@
   register: java7
-  tags: java7
+  tags:
+    - never
```
What is the use case for `never` as a tag in that context?

sxa reviewed (changes requested):
Thanks for all the responses :-)

> Actually, what I am trying to achieve here is being able to specify a username (e.g., aixtools as username) on the command-line using ansible-playbook --tags=users -e "users=aixtools" ..../main.yml because my local user for building needs the settings jenkins needs.

Understood - that makes sense. At the moment it's quite unusual for people in the admin team (other than you and I!) to create their own users instead of just using the `jenkins` user. Having said that, something like that might well make it easier to improve the process of [granting people temporary access](https://github.com/adoptium/infrastructure/blob/master/FAQ.md#temporary-access-to-a-machine) to the machines, although I think that should possibly be split out from this PR into a separate one that we can look at doing across the operating systems. Perhaps we could keep nagios and zeus users as-is and use the new parameterised `users` role for the jenkins user and any others used for build/test work. Using such a PR would be a good place to look at using a group for such users too.

> And in a proper role these would be located in roles/rolename/{vars|tasks}/ respectively.

If you look at the [Common role within the generic UNIX playbook](https://github.com/adoptium/infrastructure/tree/master/ansible/playbooks/AdoptOpenJDK_Unix_Playbook/roles/Common) it already works that way - that's the one that has the most differences across platforms

> In AIX - ulimits are ulimits, don't care what your username is.

Hmmm aren't they per-process (inherited by children)?

> as to unfying the playbooks - I have glanced at the Unix playbooks and see many opportunities for making better use of Ansible constructs AND I see the AIX playbooks as a relatively safe way to gain experience with using code

Understood, but I think if we have good ideas we can skip the "safe" option and try to do a good thing everywhere :-)

I'll catch you when you're back off vacation, but if we split out the user stuff and have a chat on the RBAC role then this is pretty much ready to merge.

aixtools commented:
> Thanks for all the responses :-)
> 
> > Actually, what I am trying to achieve here is being able to specify a username (e.g., aixtools as username) on the command-line using ansible-playbook --tags=users -e "users=aixtools" ..../main.yml because my local user for building needs the settings jenkins needs.
> 
> Understood - that makes sense. At the moment it's quite unusual for people in the admin team (other than you and I!) to create their own users instead of just using the `jenkins` user. Having said that, something like that might well make it easier to improve the process of [granting people temporary access](https://github.com/adoptium/infrastructure/blob/master/FAQ.md#temporary-access-to-a-machine) to the machines,
- I'd been taking the approach that the playbooks were being used by an _independent_ party, not directly associated with _Adoptium_.
>although I think that should possibly be split out from this PR into a separate one that we can look at doing across the operating systems. 
- Ok.
> Perhaps we could keep nagios and zeus users as-is and use the new parameterised `users` role for the jenkins user and any others used for build/test work.
- I have removed zeus and nagios from the user_list but still manage their creation/management via _users_ role (in main.yml) to keep user management, as much as possible, in one location - rather than three different directories and method (two as roles, one as include_playbook).
- Since user creation is being managed by ansible _user:_ directive I am also going to change the absolute deletion of `${HOME}/.profile` - only doing this on user creation.
- Further, want to make use of of some ansible abilities to have a template `.profile` for each user. Trying to resolve user issues in /etc/environment is a continual risk to OS management.
- **NOTE**: the .profile additions are for later!
> 
> > And in a proper role these would be located in roles/rolename/{vars|tasks}/ respectively.
> 
> If you look at the [Common role within the generic UNIX playbook](https://github.com/adoptium/infrastructure/tree/master/ansible/playbooks/AdoptOpenJDK_Unix_Playbook/roles/Common) it already works that way - that's the one that has the most differences across platforms
- Looked at this, and shall keep it in mind moving forward with bits that are currently individual roles for AIX. That is, e.g., during branch `aixpb_v2-03` I'll look at creating a _common_ role for AIX - asif - AIX was just one of the platforms being managed - as preparations for a merger.
> 
> > In AIX - ulimits are ulimits, don't care what your username is.
> 
> Hmmm aren't they per-process (inherited by children)?
- Yes, inherited - so changes made do not affect a current process, nor it's children.
- `su` and I expect `sudo` are managed differently - not just a fork/exec - but do re-evaluate the (changed) limits settings (specifically, `su - username` and `su username` both spawn with the current ulimit settings.
> 
> > as to unfying the playbooks - I have glanced at the Unix playbooks and see many opportunities for making better use of Ansible constructs AND I see the AIX playbooks as a relatively safe way to gain experience with using code
> 
> Understood, but I think if we have good ideas we can skip the "safe" option and try to do a good thing everywhere :-)
- by safe, I meant I have no chance of breaking the *Unix* playbooks
- And, it is simpler to test - as only one platform is involved.
> 
> I'll catch you when you're back off vacation, but if we split out the user stuff and have a chat on the RBAC role then this is pretty much ready to merge.
- Ok

aixtools commented on ansible/playbooks/AdoptOpenJDK_AIX_Playbook/roles/openssl/tasks/main.yml:
```diff
@@ -1,40 +1,29 @@
-  tags: openssl
+- name: "Fail: OpenSSL 1.0.2p or later required"
+  fail:
+    msg: "OpenSSL is required. Version 1.0.2p (1.0.2.1601) or later"
```
Changes made - but rather than change the _name:_ I am adding an additional _msg:_ string. I hope that meets your requirements.

aixtools commented:
Changes submitted.

sxa reviewed (approved):
Looks OK to me now :-)
(FYI @Willsparker in case you want to make any comments on the ansible-ness of this)

aixtools commented:
The next changes - mostly written - await this merger. By Friday would be great as I would hope to submit some minor and more updates in other areas this weekend.

sxa commented:
@sej-jackson @AdamBrousseau Any concerns with this before we merge it?

AdamBrousseau reviewed (commented):
I'll defer to @sej-jackson since she has been working with Michael.

sxa commented:
Not hearing much from the others so I'm merging this
